File: backend/fastapi-guide/resources/middleware-examples.py
"""
FastAPI Middleware Examples

Production-ready middleware for logging, authentication, rate limiting,
CORS, and error handling.

Usage:
    from middleware_examples import setup_middleware
    setup_middleware(app)
"""

from fastapi import FastAPI, Request, Response, status
from fastapi.middleware.cors import CORSMiddleware
from fastapi.middleware.gzip import GZipMiddleware
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from typing import Callable
import time
import logging
import structlog
from datetime import datetime


# ====================
# LOGGING MIDDLEWARE
# ====================

# Configure structured logging
logger = structlog.get_logger()


class RequestLoggingMiddleware(BaseHTTPMiddleware):
    """Log all requests and responses"""

    async def dispatch(self, request: Request, call_next: Callable):
        # Generate request ID
        request_id = f"{int(time.time() * 1000)}-{id(request)}"
        request.state.request_id = request_id

        # Log request
        logger.info(
            "request_started",
            request_id=request_id,
            method=request.method,
            path=request.url.path,
            client_host=request.client.host if request.client else None,
        )

        # Process request
        start_time = time.time()
        try:
            response = await call_next(request)
        except Exception as e:
            # Log error
            duration = time.time() - start_time
            logger.error(
                "request_failed",
                request_id=request_id,
                method=request.method,
                path=request.url.path,
                duration_ms=round(duration * 1000, 2),
                error=str(e),
                error_type=type(e).__name__,
            )
            raise

        # Log response
        duration = time.time() - start_time
        logger.info(
            "request_completed",
            request_id=request_id,
            method=request.method,
            path=request.url.path,
            status_code=response.status_code,
            duration_ms=round(duration * 1000, 2),
        )

        # Add request ID to response headers
        response.headers["X-Request-ID"] = request_id

        return response


# ====================
# RATE LIMITING MIDDLEWARE
# ====================

from collections import defaultdict
from datetime import datetime, timedelta


class RateLimitMiddleware(BaseHTTPMiddleware):
    """Simple in-memory rate limiting"""

    def __init__(self, app, requests_per_minute: int = 60):
        super().__init__(app)
        self.requests_per_minute = requests_per_minute
        self.requests = defaultdict(list)

    async def dispatch(self, request: Request, call_next: Callable):
        # Skip rate limiting for health check
        if request.url.path == "/health":
            return await call_next(request)

        # Get client identifier (IP address)
        client_id = request.client.host if request.client else "unknown"

        # Clean old requests
        now = datetime.utcnow()
        minute_ago = now - timedelta(minutes=1)
        self.requests[client_id] = [
            req_time for req_time in self.requests[client_id]
            if req_time > minute_ago
        ]

        # Check rate limit
        if len(self.requests[client_id]) >= self.requests_per_minute:
            return JSONResponse(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                content={
                    "error": "Rate limit exceeded",
                    "detail": f"Maximum {self.requests_per_minute} requests per minute"
                }
            )

        # Record request
        self.requests[client_id].append(now)

        return await call_next(request)


# ====================
# AUTHENTICATION MIDDLEWARE
# ====================

class AuthenticationMiddleware(BaseHTTPMiddleware):
    """Add user info to request state"""

    async def dispatch(self, request: Request, call_next: Callable):
        # Public endpoints that don't require auth
        public_paths = ["/health", "/docs", "/redoc", "/openapi.json", "/auth/login", "/auth/register"]

        if any(request.url.path.startswith(path) for path in public_paths):
            return await call_next(request)

        # Get token from header
        auth_header = request.headers.get("Authorization")

        if not auth_header or not auth_header.startswith("Bearer "):
            return JSONResponse(
                status_code=status.HTTP_401_UNAUTHORIZED,
                content={"error": "Authentication required"},
                headers={"WWW-Authenticate": "Bearer"},
            )

        token = auth_header.split(" ")[1]

        try:
            # Verify token (implement your JWT verification)
            # This is a placeholder
            user_id = verify_token(token)
            request.state.user_id = user_id
        except Exception as e:
            return JSONResponse(
                status_code=status.HTTP_401_UNAUTHORIZED,
                content={"error": "Invalid or expired token"},
                headers={"WWW-Authenticate": "Bearer"},
            )

        return await call_next(request)


# ====================
# SECURITY HEADERS MIDDLEWARE
# ====================

class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """Add security headers to all responses"""

    async def dispatch(self, request: Request, call_next: Callable):
        response = await call_next(request)

        # Add security headers
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-XSS-Protection"] = "1; mode=block"
        response.headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"

        return response


# ====================
# TIMING MIDDLEWARE
# ====================

class TimingMiddleware(BaseHTTPMiddleware):
    """Add Server-Timing header for performance monitoring"""

    async def dispatch(self, request: Request, call_next: Callable):
        start_time = time.time()
        response = await call_next(request)
        duration = time.time() - start_time

        # Add timing header
        response.headers["Server-Timing"] = f"total;dur={duration * 1000:.2f}"

        return response


# ====================
# ERROR HANDLING MIDDLEWARE
# ====================

from fastapi.exceptions import RequestValidationError
from pydantic import ValidationError


async def validation_exception_handler(request: Request, exc: RequestValidationError):
    """Handle Pydantic validation errors"""
    errors = []
    for error in exc.errors():
        errors.append({
            "field": ".".join(str(loc) for loc in error["loc"]),
            "message": error["msg"],
            "type": error["type"],
        })

    return JSONResponse(
        status_code=status.HTTP_400_BAD_REQUEST,
        content={
            "error": "Validation failed",
            "details": errors,
            "timestamp": datetime.utcnow().isoformat(),
        }
    )


async def generic_exception_handler(request: Request, exc: Exception):
    """Handle all other exceptions"""
    # Log the error
    logger.error(
        "unhandled_exception",
        error=str(exc),
        error_type=type(exc).__name__,
        path=request.url.path,
        method=request.method,
    )

    # Don't expose internal errors in production
    from app.core.config import settings
    detail = str(exc) if settings.DEBUG else "Internal server error"

    return JSONResponse(
        status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
        content={
            "error": detail,
            "timestamp": datetime.utcnow().isoformat(),
        }
    )


# ====================
# MAINTENANCE MODE MIDDLEWARE
# ====================

class MaintenanceModeMiddleware(BaseHTTPMiddleware):
    """Enable maintenance mode via environment variable"""

    async def dispatch(self, request: Request, call_next: Callable):
        from app.core.config import settings

        # Allow health check during maintenance
        if request.url.path == "/health":
            return await call_next(request)

        if settings.MAINTENANCE_MODE:
            return JSONResponse(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                content={
                    "error": "Service under maintenance",
                    "message": settings.MAINTENANCE_MESSAGE or "We'll be back soon!",
                }
            )

        return await call_next(request)


# ====================
# SETUP FUNCTION
# ====================

def setup_middleware(app: FastAPI):
    """
    Set up all middleware for the FastAPI application

    Call this function after creating your FastAPI app:
        app = FastAPI()
        setup_middleware(app)
    """

    # CORS (must be first)
    app.add_middleware(
        CORSMiddleware,
        allow_origins=["*"],  # Configure based on settings
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )

    # Compression
    app.add_middleware(GZipMiddleware, minimum_size=1000)

    # Custom middleware (order matters)
    app.add_middleware(MaintenanceModeMiddleware)
    app.add_middleware(SecurityHeadersMiddleware)
    app.add_middleware(RateLimitMiddleware, requests_per_minute=100)
    app.add_middleware(TimingMiddleware)
    app.add_middleware(RequestLoggingMiddleware)

    # Exception handlers
    app.add_exception_handler(RequestValidationError, validation_exception_handler)
    app.add_exception_handler(Exception, generic_exception_handler)

    logging.info("Middleware setup complete")


# ====================
# UTILITY FUNCTIONS
# ====================

def verify_token(token: str) -> int:
    """
    Verify JWT token and return user ID

    This is a placeholder - implement with your JWT library
    """
    import jwt
    from app.core.config import settings

    try:
        payload = jwt.decode(token, settings.SECRET_KEY, algorithms=["HS256"])
        user_id = payload.get("sub")
        if user_id is None:
            raise ValueError("Invalid token payload")
        return int(user_id)
    except jwt.ExpiredSignatureError:
        raise ValueError("Token has expired")
    except jwt.InvalidTokenError:
        raise ValueError("Invalid token")


# ====================
# EXAMPLE USAGE
# ====================

"""
# main.py

from fastapi import FastAPI
from middleware_examples import setup_middleware
from routes import users, auth

app = FastAPI(
    title="My API",
    description="API with comprehensive middleware",
    version="1.0.0"
)

# Setup middleware
setup_middleware(app)

# Include routers
app.include_router(users.router)
app.include_router(auth.router)

# Health check
@app.get("/health")
async def health():
    return {
        "status": "ok",
        "timestamp": datetime.utcnow().isoformat()
    }
"""

File: backend/nodejs-api-guide/resources/middleware-examples.ts
/**
 * Express Middleware Examples
 *
 * Production-ready middleware for Node.js REST APIs including:
 * - Authentication (JWT)
 * - Authorization (role-based, resource-based)
 * - Error handling
 * - Request logging
 * - Rate limiting
 * - CORS configuration
 * - Request ID tracking
 * - Response compression
 *
 * @version 1.0.0
 */

import { Request, Response, NextFunction } from 'express';
import jwt from 'jsonwebtoken';
import rateLimit from 'express-rate-limit';
import helmet from 'helmet';
import cors from 'cors';
import compression from 'compression';
import { randomUUID } from 'crypto';
import pino from 'pino';

// ============================================================================
// Type Definitions
// ============================================================================

/**
 * Extended Request with user and requestId
 */
export interface AuthRequest extends Request {
  user?: {
    userId: string;
    email: string;
    role: string;
  };
  requestId?: string;
}

/**
 * Custom error class with status code
 */
export class AppError extends Error {
  constructor(
    public statusCode: number,
    public message: string,
    public isOperational = true,
    public details?: any
  ) {
    super(message);
    Object.setPrototypeOf(this, AppError.prototype);
    Error.captureStackTrace(this, this.constructor);
  }
}

// ============================================================================
// Logger Configuration
// ============================================================================

export const logger = pino({
  level: process.env.LOG_LEVEL || 'info',
  transport:
    process.env.NODE_ENV === 'development'
      ? {
          target: 'pino-pretty',
          options: {
            colorize: true,
            translateTime: 'SYS:standard',
            ignore: 'pid,hostname',
          },
        }
      : undefined,
  redact: {
    paths: [
      'req.headers.authorization',
      'req.headers.cookie',
      'password',
      'token',
      'accessToken',
      'refreshToken',
    ],
    remove: true,
  },
});

// ============================================================================
// Request ID Middleware
// ============================================================================

/**
 * Add unique request ID to every request
 * Useful for tracing requests through logs
 */
export function requestId(req: AuthRequest, res: Response, next: NextFunction) {
  req.requestId = req.headers['x-request-id'] as string || randomUUID();
  res.setHeader('X-Request-ID', req.requestId);
  next();
}

// ============================================================================
// Request Logging Middleware
// ============================================================================

/**
 * Log all incoming requests
 */
export function requestLogger(req: AuthRequest, res: Response, next: NextFunction) {
  const start = Date.now();

  // Log request
  logger.info({
    requestId: req.requestId,
    method: req.method,
    url: req.url,
    ip: req.ip,
    userAgent: req.headers['user-agent'],
  }, 'Incoming request');

  // Log response when finished
  res.on('finish', () => {
    const duration = Date.now() - start;

    logger.info({
      requestId: req.requestId,
      method: req.method,
      url: req.url,
      statusCode: res.statusCode,
      duration: `${duration}ms`,
    }, 'Request completed');
  });

  next();
}

// ============================================================================
// Authentication Middleware
// ============================================================================

/**
 * Verify JWT token from Authorization header
 * Throws 401 if token is missing or invalid
 */
export function authenticate(req: AuthRequest, res: Response, next: NextFunction) {
  const authHeader = req.headers['authorization'];

  if (!authHeader) {
    return next(new AppError(401, 'Authentication required', true, {
      code: 'NO_TOKEN',
    }));
  }

  const parts = authHeader.split(' ');

  if (parts.length !== 2 || parts[0] !== 'Bearer') {
    return next(new AppError(401, 'Invalid authorization format', true, {
      code: 'INVALID_FORMAT',
    }));
  }

  const token = parts[1];

  try {
    const decoded = jwt.verify(token, process.env.JWT_SECRET!) as any;
    req.user = {
      userId: decoded.userId,
      email: decoded.email,
      role: decoded.role,
    };

    logger.debug({
      requestId: req.requestId,
      userId: req.user.userId,
    }, 'User authenticated');

    next();
  } catch (error) {
    if (error instanceof jwt.TokenExpiredError) {
      return next(new AppError(401, 'Token expired', true, {
        code: 'TOKEN_EXPIRED',
      }));
    }

    if (error instanceof jwt.JsonWebTokenError) {
      return next(new AppError(401, 'Invalid token', true, {
        code: 'INVALID_TOKEN',
      }));
    }

    next(error);
  }
}

/**
 * Optional authentication middleware
 * Doesn't fail if no token provided, but validates if present
 */
export function optionalAuth(req: AuthRequest, res: Response, next: NextFunction) {
  const authHeader = req.headers['authorization'];

  if (!authHeader) {
    return next();
  }

  const parts = authHeader.split(' ');

  if (parts.length === 2 && parts[0] === 'Bearer') {
    const token = parts[1];

    try {
      const decoded = jwt.verify(token, process.env.JWT_SECRET!) as any;
      req.user = {
        userId: decoded.userId,
        email: decoded.email,
        role: decoded.role,
      };
    } catch (error) {
      // Silently fail for optional auth
      logger.debug({ error }, 'Optional auth failed');
    }
  }

  next();
}

// ============================================================================
// Authorization Middleware
// ============================================================================

/**
 * Authorize specific roles
 * Must be used after authenticate middleware
 */
export function authorize(...allowedRoles: string[]) {
  return (req: AuthRequest, res: Response, next: NextFunction) => {
    if (!req.user) {
      return next(new AppError(401, 'Authentication required', true, {
        code: 'NO_AUTH',
      }));
    }

    if (!allowedRoles.includes(req.user.role)) {
      logger.warn({
        requestId: req.requestId,
        userId: req.user.userId,
        role: req.user.role,
        allowedRoles,
        url: req.url,
      }, 'Authorization failed');

      return next(new AppError(403, 'Insufficient permissions', true, {
        code: 'FORBIDDEN',
      }));
    }

    next();
  };
}

/**
 * Verify user owns the resource
 * Checks if :userId param matches authenticated user
 */
export function authorizeOwnership(req: AuthRequest, res: Response, next: NextFunction) {
  if (!req.user) {
    return next(new AppError(401, 'Authentication required'));
  }

  const resourceUserId = req.params.userId || req.body.userId;

  if (!resourceUserId) {
    return next(new AppError(400, 'User ID not found in request'));
  }

  if (req.user.userId !== resourceUserId && req.user.role !== 'admin') {
    logger.warn({
      requestId: req.requestId,
      userId: req.user.userId,
      resourceUserId,
    }, 'Ownership authorization failed');

    return next(new AppError(403, 'You do not have permission to access this resource'));
  }

  next();
}

/**
 * Check if user is admin or owner
 */
export function authorizeAdminOrOwner(userIdField = 'userId') {
  return (req: AuthRequest, res: Response, next: NextFunction) => {
    if (!req.user) {
      return next(new AppError(401, 'Authentication required'));
    }

    const resourceUserId = req.params[userIdField] || req.body[userIdField];

    if (req.user.role === 'admin' || req.user.userId === resourceUserId) {
      return next();
    }

    return next(new AppError(403, 'Insufficient permissions'));
  };
}

// ============================================================================
// Error Handling Middleware
// ============================================================================

/**
 * Not found handler (404)
 * Should be placed after all routes
 */
export function notFoundHandler(req: Request, res: Response, next: NextFunction) {
  next(new AppError(404, `Route ${req.method} ${req.url} not found`, true, {
    code: 'NOT_FOUND',
  }));
}

/**
 * Global error handler
 * Should be placed last in middleware chain
 */
export function errorHandler(
  err: Error | AppError,
  req: AuthRequest,
  res: Response,
  next: NextFunction
) {
  // Default to 500 server error
  let statusCode = 500;
  let message = 'Internal server error';
  let details: any = undefined;

  // Handle known operational errors
  if (err instanceof AppError) {
    statusCode = err.statusCode;
    message = err.message;
    details = err.details;

    // Log operational errors as warnings
    logger.warn({
      requestId: req.requestId,
      error: message,
      statusCode,
      details,
      url: req.url,
      userId: req.user?.userId,
    }, 'Operational error');
  } else {
    // Log unexpected errors
    logger.error({
      requestId: req.requestId,
      error: err.message,
      stack: err.stack,
      url: req.url,
      userId: req.user?.userId,
    }, 'Unexpected error');
  }

  // Don't expose stack traces in production
  const isDevelopment = process.env.NODE_ENV === 'development';

  res.status(statusCode).json({
    error: message,
    ...(details && { details }),
    ...(isDevelopment && { stack: err.stack }),
    requestId: req.requestId,
  });
}

/**
 * Async error wrapper
 * Catches async errors and passes to error handler
 */
export function asyncHandler(fn: Function) {
  return (req: Request, res: Response, next: NextFunction) => {
    Promise.resolve(fn(req, res, next)).catch(next);
  };
}

// ============================================================================
// Rate Limiting Middleware
// ============================================================================

/**
 * General API rate limiter
 */
export const generalLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 100, // Limit each IP to 100 requests per windowMs
  message: 'Too many requests from this IP, please try again later.',
  standardHeaders: true,
  legacyHeaders: false,
  handler: (req, res) => {
    logger.warn({
      ip: req.ip,
      url: req.url,
    }, 'Rate limit exceeded');

    res.status(429).json({
      error: 'Too many requests, please try again later.',
      retryAfter: res.getHeader('Retry-After'),
    });
  },
});

/**
 * Strict rate limiter for authentication endpoints
 */
export const authLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 5, // 5 attempts
  skipSuccessfulRequests: true,
  message: 'Too many authentication attempts, please try again later.',
  standardHeaders: true,
  legacyHeaders: false,
});

/**
 * API key rate limiter (more generous)
 */
export const apiKeyLimiter = rateLimit({
  windowMs: 60 * 1000, // 1 minute
  max: 60, // 60 requests per minute
  message: 'API rate limit exceeded.',
});

/**
 * File upload rate limiter
 */
export const uploadLimiter = rateLimit({
  windowMs: 60 * 60 * 1000, // 1 hour
  max: 10, // 10 uploads per hour
  message: 'Too many uploads, please try again later.',
});

// ============================================================================
// CORS Configuration
// ============================================================================

/**
 * CORS middleware with environment-based configuration
 */
export const corsMiddleware = cors({
  origin: (origin, callback) => {
    const allowedOrigins = process.env.ALLOWED_ORIGINS?.split(',') || ['http://localhost:3000'];

    // Allow requests with no origin (like mobile apps or curl)
    if (!origin) return callback(null, true);

    if (allowedOrigins.includes(origin)) {
      callback(null, true);
    } else {
      logger.warn({ origin }, 'CORS request blocked');
      callback(new Error('Not allowed by CORS'));
    }
  },
  credentials: true,
  methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS'],
  allowedHeaders: ['Content-Type', 'Authorization', 'X-Request-ID'],
  exposedHeaders: ['X-Request-ID'],
  maxAge: 86400, // 24 hours
});

// ============================================================================
// Security Headers
// ============================================================================

/**
 * Security headers middleware using Helmet
 */
export const securityHeaders = helmet({
  contentSecurityPolicy: {
    directives: {
      defaultSrc: ["'self'"],
      scriptSrc: ["'self'"],
      styleSrc: ["'self'", "'unsafe-inline'"],
      imgSrc: ["'self'", "data:", "https:"],
      connectSrc: ["'self'"],
      fontSrc: ["'self'"],
      objectSrc: ["'none'"],
      mediaSrc: ["'self'"],
      frameSrc: ["'none'"],
    },
  },
  hsts: {
    maxAge: 31536000,
    includeSubDomains: true,
    preload: true,
  },
  noSniff: true,
  xssFilter: true,
  hidePoweredBy: true,
});

// ============================================================================
// Compression Middleware
// ============================================================================

/**
 * Response compression middleware
 */
export const compressionMiddleware = compression({
  filter: (req, res) => {
    if (req.headers['x-no-compression']) {
      return false;
    }
    return compression.filter(req, res);
  },
  level: 6, // Balance between speed and compression ratio
});

// ============================================================================
// Request Validation Helpers
// ============================================================================

/**
 * Validate request content type
 */
export function requireContentType(...allowedTypes: string[]) {
  return (req: Request, res: Response, next: NextFunction) => {
    const contentType = req.headers['content-type'];

    if (!contentType) {
      return next(new AppError(400, 'Content-Type header is required'));
    }

    const isAllowed = allowedTypes.some((type) => contentType.includes(type));

    if (!isAllowed) {
      return next(new AppError(415, `Unsupported media type. Expected: ${allowedTypes.join(', ')}`));
    }

    next();
  };
}

/**
 * Require JSON content type
 */
export const requireJSON = requireContentType('application/json');

// ============================================================================
// Pagination Middleware
// ============================================================================

/**
 * Parse and validate pagination parameters
 */
export function parsePagination(req: AuthRequest, res: Response, next: NextFunction) {
  const page = Math.max(1, parseInt(req.query.page as string) || 1);
  const limit = Math.min(100, Math.max(1, parseInt(req.query.limit as string) || 20));
  const offset = (page - 1) * limit;

  (req as any).pagination = {
    page,
    limit,
    offset,
  };

  next();
}

// ============================================================================
// Response Helpers
// ============================================================================

/**
 * Add helper methods to response object
 */
export function responseHelpers(req: Request, res: Response, next: NextFunction) {
  // Success response
  (res as any).success = (data: any, message?: string) => {
    res.json({
      success: true,
      ...(message && { message }),
      data,
    });
  };

  // Paginated response
  (res as any).paginated = (data: any[], total: number, page: number, limit: number) => {
    const totalPages = Math.ceil(total / limit);

    res.json({
      success: true,
      data,
      pagination: {
        page,
        limit,
        total,
        totalPages,
        hasNextPage: page < totalPages,
        hasPrevPage: page > 1,
      },
    });
  };

  // Error response
  (res as any).error = (statusCode: number, message: string, details?: any) => {
    res.status(statusCode).json({
      success: false,
      error: message,
      ...(details && { details }),
    });
  };

  next();
}

// ============================================================================
// API Version Middleware
// ============================================================================

/**
 * Check API version from header or URL
 */
export function apiVersion(supportedVersions: string[]) {
  return (req: Request, res: Response, next: NextFunction) => {
    const versionHeader = req.headers['api-version'] as string;
    const versionUrl = req.path.match(/^\/v(\d+)\//)?.[1];

    const version = versionHeader || versionUrl || '1';

    if (!supportedVersions.includes(version)) {
      return next(new AppError(400, `API version ${version} is not supported`, true, {
        supportedVersions,
      }));
    }

    (req as any).apiVersion = version;
    next();
  };
}

// ============================================================================
// Maintenance Mode Middleware
// ============================================================================

/**
 * Return 503 when in maintenance mode
 */
export function maintenanceMode(req: Request, res: Response, next: NextFunction) {
  if (process.env.MAINTENANCE_MODE === 'true') {
    return res.status(503).json({
      error: 'Service temporarily unavailable for maintenance',
      message: process.env.MAINTENANCE_MESSAGE || 'We\'ll be back soon!',
    });
  }
  next();
}

// ============================================================================
// Usage Example
// ============================================================================

/**
 * Example Express app setup:
 *
 * import express from 'express';
 * import * as middleware from './middleware-examples';
 *
 * const app = express();
 *
 * // Security & basics
 * app.use(middleware.securityHeaders);
 * app.use(middleware.corsMiddleware);
 * app.use(middleware.compressionMiddleware);
 * app.use(express.json({ limit: '10mb' }));
 * app.use(express.urlencoded({ extended: true }));
 *
 * // Request tracking
 * app.use(middleware.requestId);
 * app.use(middleware.requestLogger);
 *
 * // Response helpers
 * app.use(middleware.responseHelpers);
 *
 * // Maintenance mode check
 * app.use(middleware.maintenanceMode);
 *
 * // Rate limiting
 * app.use('/api/', middleware.generalLimiter);
 * app.use('/api/auth/', middleware.authLimiter);
 *
 * // Public routes
 * app.post('/api/auth/login', authController.login);
 *
 * // Protected routes
 * app.get('/api/users',
 *   middleware.authenticate,
 *   middleware.parsePagination,
 *   userController.list
 * );
 *
 * // Admin routes
 * app.delete('/api/users/:id',
 *   middleware.authenticate,
 *   middleware.authorize('admin'),
 *   userController.delete
 * );
 *
 * // Owner or admin routes
 * app.put('/api/users/:userId/profile',
 *   middleware.authenticate,
 *   middleware.authorizeAdminOrOwner('userId'),
 *   userController.updateProfile
 * );
 *
 * // Error handlers (must be last)
 * app.use(middleware.notFoundHandler);
 * app.use(middleware.errorHandler);
 *
 * export default app;
 */
